File: leetcode/cpp/433. 最小基因变化.cpp
class Solution {
public:
    int minMutation(string start, string end, vector<string>& bank) {
        // 初始化
        int step = 0, size;
        string letters = "ACGT", currWord;
        queue<string> lqueue, rqueue;
        lqueue.push(start), rqueue.push(end);
        unordered_set<string> lvisited, rvisited, wordSet(bank.begin(), bank.end());
        lvisited.insert(start), rvisited.insert(end);

        // Double-BFS
        while (!lqueue.empty() && !rqueue.empty()) {
            if (lqueue.size() > rqueue.size()) {
                swap(lqueue, rqueue);
                swap(lvisited, rvisited);
            }
            size = lqueue.size();
            for (int i = 0; i < size; i++) {
                currWord = lqueue.front();
                lqueue.pop();
                if (rvisited.count(currWord) > 0) {
                    return step;
                }
                for (int j = 0; j < currWord.size(); j++) {
                    char oriChar = currWord[j];
                    for (const char & newChar : letters) {
                        if (newChar != currWord[j]) {
                            currWord[j] = newChar;
                            if (lvisited.count(currWord) == 0 && wordSet.count(currWord) > 0) {
                                lqueue.push(currWord);
                                lvisited.insert(currWord);
                            }
                        }
                    }
                    currWord[j] = oriChar;
                }
            }
            step += 1;
        }
        return -1;
    }
};

File: leetcode/python/433. 最小基因变化.py
#!/bin/bash python
"""
433. 最小基因变化

思路：
双向BFS，与127同
"""

class Solution(object):
    def minMutation(self, start, end, bank):
        """
        :type start: str
        :type end: str
        :type bank: List[str]
        :rtype: int
        与127单词接龙基本上是一道题，双端BFS
        """
        # 初始化
        step = 0
        letters = ["A", "C", "G", "T"]
        lqueue, rqueue = deque(), deque()
        lqueue.append(start)
        rqueue.append(end)
        lvisited, rvisited, wordSet = set(), set(), set(bank)
        lvisited.add(start)
        rvisited.add(end)

        # Double-BFS
        while lqueue and rqueue:
            if len(lqueue) > len(rqueue):
                lqueue, rqueue = rqueue, lqueue
                lvisited, rvisited = rvisited, lvisited
            size = len(lqueue)
            for i in range(size):
                currWord = lqueue.popleft()
                if currWord in rvisited:
                    return step
                for j in range(len(currWord)):
                    for char in letters:
                        nextWord = currWord[:j] + char + currWord[j+1:]
                        if nextWord not in lvisited and nextWord in wordSet:
                            lvisited.add(nextWord)
                            lqueue.append(nextWord)
            step += 1
        return -1